### PR TITLE
Add Kotlin to build

### PR DESCRIPTION
**Pull request by seanf**
https://zanata.atlassian.net/browse/ZNTA-1661

To investigate:

- [x] ~~https://github.com/shyiko/ktlint~~ Going with detekt.
- [x] https://github.com/arturbosch/detekt Waiting for next release.
- [x] ~~Groovy script for Maven to ensure `.kt` files go under `src/*/kotlin`, not `src/*/java`. Ref: [IDEA-166978](https://youtrack.jetbrains.com/issue/IDEA-166978)~~ - now using `src/*/java` as recommended by JetBrains.

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-server/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-server/wiki/Development-Guidelines#new-technologies-and-dependencies)?

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/136)

### CODE REVIEW

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by seanf** [comment]
@zanata-jenkins test this please

File: server/zanata-model/src/main/kotlin/org/zanata/model/HLocale.kt [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    @get:NotNull
+    @get:Type(type = "localeId")
+    var localeId: LocaleId
+
+    var isActive: Boolean = false
+
+    var isEnabledByDefault: Boolean = false
+    var supportedProjects: Set<HProject>? = null
+        @ManyToMany
+        @JoinTable(name = "HProject_Locale", joinColumns = arrayOf(JoinColumn(name = "localeId")), inverseJoinColumns = arrayOf(JoinColumn(name = "projectId")))
```
Comment (aeng) [inline server/zanata-model/src/main/kotlin/org/zanata/model/HLocale.kt]: Any checkstyle rules for max-length of a line?

File: server/zanata-model/src/main/kotlin/org/zanata/model/HLocale.kt [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        get() {
+            if (field == null)
+                this.supportedProjects = HashSet<HProject>()
+            return field
+        }
+    var supportedIterations: Set<HProjectIteration>? = null
+        @ManyToMany
+        @JoinTable(name = "HProjectIteration_Locale", joinColumns = arrayOf(JoinColumn(name = "localeId")), inverseJoinColumns = arrayOf(JoinColumn(name = "projectIterationId")))
```
Comment (aeng) [inline server/zanata-model/src/main/kotlin/org/zanata/model/HLocale.kt]: Any checkstyle rules for max-length of a line?

File: server/zanata-war/src/main/kotlin/org/zanata/dao/LocaleDAO.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import com.google.common.collect.Lists
+
+@RequestScoped
+class LocaleDAO : AbstractDAOImpl<HLocale, Long> {
+
+    constructor() : super(HLocale::class.java) {}
+
+    constructor(session: Session) : super(HLocale::class.java, session) {}
+
+    fun findByLocaleId(locale: LocaleId): HLocale? {
```
Comment (aeng) [inline server/zanata-war/src/main/kotlin/org/zanata/dao/LocaleDAO.kt]: What the ? behind HLocale meant? nullable?

File: server/zanata-war/src/main/kotlin/org/zanata/adapter/TranslatableSeparator.kt [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        val stripAllNonTranslatable = Function<SplitString, SplitString> { stripTrailingEmptyPairedTags(it) }
+                .compose(Function<SplitString, SplitString> { stripLeadingEmptyPairedTags(it) })
+                .compose(Function<SplitString, SplitString> { stripWrappingTags(it) })
+                .compose(Function<SplitString, SplitString> { stripTrailingEmptyTag(it) })
+                .compose(Function<SplitString, SplitString> { stripLeadingEmptyTags(it) })
+                .compose(Function<SplitString, SplitString> { stripTrailingWhitespace(it) })
+                .compose(Function<SplitString, SplitString> { stripLeadingWhitespace(it) })
+        var preStripped = SplitString("", s, "")
+        var stripped = stripAllNonTranslatable.apply(preStripped)
+        while (preStripped != stripped) {
```
Comment (aeng) [inline server/zanata-war/src/main/kotlin/org/zanata/adapter/TranslatableSeparator.kt]: Does Kotlin interpret `==` as `.equal` in Java?

File: server/zanata-war/src/main/kotlin/org/zanata/dao/LocaleDAO.kt (hunk 3 above)
Comment (seanf) [inline server/zanata-war/src/main/kotlin/org/zanata/dao/LocaleDAO.kt]: Yes, nullable. Without `?`, it's not nullable.

File: server/zanata-war/src/main/kotlin/org/zanata/adapter/TranslatableSeparator.kt (hunk 4 above)
Comment (seanf) [inline server/zanata-war/src/main/kotlin/org/zanata/adapter/TranslatableSeparator.kt]: Yep. `===` for checking identity.

File: server/pom.xml [hunk 5]
```diff
@@ -2004,7 +2018,82 @@
...
+            <artifactId>detekt-cli</artifactId>
+            <version>1.0.0.M6</version>
+          </dependency>
+          <dependency>
+            <groupId>io.gitlab.arturbosch.detekt</groupId>
+            <artifactId>detekt-formatting</artifactId>
+            <version>1.0.0.M6</version>
+          </dependency>
+<!--
+          <dependency>
```
Comment (aeng) [inline server/pom.xml]: Should remove this if its not used anymore.

Comment (aeng) [inline server/pom.xml]: Should remove this block if not used anymore.

Comment (seanf) [inline server/pom.xml]: Yes, I removed it in a later commit.

**Comment by djansen-redhat** [comment]
🙈

**Pull request merged**